onflow/flow-go
# [Verification] Implements fetcher and requester engines

This PR implements the `fetcher` and `requester` engines of the verification node. 

**Overview**:
The new architecture of the verification node composed of three main engines, `assigner`, `fetcher`, and `verifier`. We borrow `verifier` intact from the current architecture. And have implemented the rest two engines. This PR is completely around the `fetcher` engine. 

**How does the verification pipeline work?**:
In a very nutshell: 
- The `assigner` extracts results from finalized blocks, performs the chunk assignment, and passes the assigned chunks to the `fetcher` engine (through a `ChunkConsumer`), i.e., `assigner` -> `ChunkConsumer` -> `fetcher`. 

- The `fetcher` receives assigned chunks from `ChunkConsumer` and requests their chunk data pack from the execution nodes. Once it has the chunk data pack of an assigned chunk, it forms a verifiable chunk and passes it to the `verifier` engine for verification. Once the verification is done and `verifier` engine returns, `fetcher` notifies `ChunkConsumer` that it is done with verifying this chunk (either successful or unsuccessful), and so the `ChunkConsumer` can dispatch the next chunk to `fetcher`. 

- Requesting chunk data packs from execution nodes by `fetcher` engine is done through a brand new 😎  engine of verification node that is called `requester`. It is responsible for receiving a chunk data request from `fetcher`, dispatching it in the network to the execution node, and receiving the chunk data pack and relaying it back to the `fetcher`. It keeps retrying pending chunk requests until it receives the chunk data pack. 

**Note** that the chunk consumer runs `workers` (i.e., handler method) of `fetcher` engine concurrently, so processing chunks at `fetcher` engine is subject to some adaptive degree of concurrency (which has been addressed previously and is not the topic of this PR). 

**Chunks of sealed blocks:** before making a chunk data request to `requester`, the `fetcher` engine makes sure that the chunk does not belong to a sealed block. Also, upon each retry at the `requester` side, the `requester` engine makes sure that the block of the chunk has not yet sealed, and drops the chunk request otherwise. 

**We have optimizations on the requester on the way:** following an incremental build strategy, the current `requester` is an advanced version of what we have currently at the `match` engine. However, we are going to implement a specific [exponential backoff mechanism](https://github.com/dapperlabs/flow-go/issues/5428) for the requester as the next step. We are also going to investigate filling the gaps between the generic requester engine, and what we currently have, [seeking the possibility to migrate the logic to the generic requester](https://github.com/dapperlabs/flow-go/issues/5419). Though it needs some refactoring (e.g., requests should be validated against sealed blocks before each dispatch, requests should be dispatched over `Publish` instead of the current `Unicast` approach, and most importantly, we must refactor the current `provider` engine at the Execution Node to utilize `common/provider` engine).
